### ISSUE TITLE
service discovery request hang-up caused by consul nodes unhealthy issue

### ISSUE THREAD

**Issue by chenyan71** [networknt/light-4j#726]
When consul cluster restart or some nodes in consul cluster in unhealthy status, the services face service discovery request hang-up and take long time to return empty url:

java.lang.NullPointerException: null 
at java.net.URI$Parser.parse(URI.java:3042) 
at java.net.URI.<init>(URI.java:588) 

After service restart, it will work fine.

The majorly issue is in the ConsulRegistry, in some cases the serviceCache doesn't not include the specified service urls since the consul issue, but the index has value already. This triggered the consul blocking queries for service discovery:

https://www.consul.io/api-docs/features/blocking

Basically there two issues may need fix:

1 . for service discovery, it should always set index as 0L to avoid consul blocking queries .
2. as consul document mentioned, if the consul response index less than previous value, reset the index to 0:

Implementations must check to see if a returned index is lower than the previous value, and if it is, should reset index to 0

**Comment by stevehu** [comment]
A Consul cluster has several nodes and they are synced constantly. However, when the network is not stable, the data might not be consistent on all nodes. This might be the reason why the Consul returned index is smaller than the cached index. I am wondering if we can reproduce the situation by a random shutdown of the Consul nodes.

**Comment by stevehu** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [service discovery request hang-up caused by consul nodes unhealthy issue 1.6.x](https://trello.com/c/0X3IjLSm/704-service-discovery-request-hang-up-caused-by-consul-nodes-unhealthy-issue-16x)

**Comment by stevehu** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [service discovery request hang-up caused by consul nodes unhealthy issue master](https://trello.com/c/z26ZkqGQ/705-service-discovery-request-hang-up-caused-by-consul-nodes-unhealthy-issue-master)
